File: langs/csharp/src/VMx/Commands/ConfirmationDecoratorCommand.cs
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Runtime.ExceptionServices;
using System.Windows.Input;

namespace VMx.Commands;

/// <summary>
/// Wraps a single inner <see cref="ICommand"/> with an async confirmation gate.
/// See spec/04-commands.md §Decorators and ADR-0012.
///
/// The confirm delegate is intentionally UI-agnostic: it returns a Task&lt;bool&gt;
/// that resolves to true if the user accepted, false otherwise. Bridge helpers
/// to a notification hub live in the optional VMx.Notifications assembly.
/// </summary>
public sealed class ConfirmationDecoratorCommand : ICommand, IDisposable
{
    private readonly ICommand _inner;
    private readonly Func<Task<bool>> _confirm;
    private readonly EventHandler _innerHandler;
    private readonly Subject<Exception> _errors = new();
    private readonly object _errorGate = new();
    private volatile bool _disposed;

    /// <summary>Creates a new <see cref="ConfirmationDecoratorCommand"/>.</summary>
    public ConfirmationDecoratorCommand(ICommand inner, Func<Task<bool>> confirm)
    {
        _inner = inner ?? throw new ArgumentNullException(nameof(inner));
        _confirm = confirm ?? throw new ArgumentNullException(nameof(confirm));
        _innerHandler = (sender, args) => CanExecuteChanged?.Invoke(this, args);
        _inner.CanExecuteChanged += _innerHandler;
    }

    /// <inheritdoc/>
    public event EventHandler? CanExecuteChanged;

    /// <summary>
    /// Observable that surfaces an error from the fire-and-forget <see cref="Execute"/>
    /// path — either the confirm delegate faulting or the inner command throwing.
    /// <see cref="ICommand.Execute"/> is <c>void</c>, so it cannot propagate the
    /// failure across the async confirm gate the way <see cref="RelayCommand"/>'s
    /// task does; instead of swallowing it (a discarded faulted <see cref="Task"/>
    /// would otherwise surface only as an <c>UnobservedTaskException</c> at GC) the
    /// error is emitted here (VMX-009). The awaitable <see cref="ExecuteAsync"/>
    /// path keeps its throw behavior — await it directly to handle the error
    /// inline. Completes on <see cref="Dispose"/>.
    /// </summary>
    public IObservable<Exception> Errors => _errors.AsObservable();

    /// <inheritdoc/>
    public bool CanExecute(object? parameter) => _inner.CanExecute(parameter);

    /// <inheritdoc/>
    /// <remarks>
    /// Synchronous <c>Execute</c> kicks off the confirmation flow and returns immediately.
    /// A rejecting confirm delegate or a throwing inner command is routed to
    /// <see cref="Errors"/> rather than swallowed (VMX-009). Use
    /// <see cref="ExecuteAsync"/> to await the full sequence and observe the failure inline.
    /// </remarks>
    public void Execute(object? parameter) =>
        _ = ExecuteAsync(parameter).ContinueWith(
            t =>
            {
                EmitError(t.Exception!.GetBaseException());
            },
            CancellationToken.None,
            TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously,
            TaskScheduler.Default);

    /// <summary>Awaits the confirm delegate, then invokes inner.Execute if accepted.</summary>
    public async Task ExecuteAsync(object? parameter)
    {
        if (!CanExecute(parameter)) return;
        var confirmed = await _confirm().ConfigureAwait(false);
        if (confirmed) _inner.Execute(parameter);
    }

    private void EmitError(Exception error)
    {
        lock (_errorGate)
        {
            if (_disposed) return;
            _errors.OnNext(error);
        }
    }

    /// <summary>
    /// Unsubscribes from the inner command's <c>CanExecuteChanged</c> and completes
    /// the <see cref="Errors"/> channel. Idempotent.
    /// </summary>
    public void Dispose()
    {
        lock (_errorGate)
        {
            if (_disposed) return;
            _disposed = true;
            _inner.CanExecuteChanged -= _innerHandler;
        }

        ExceptionDispatchInfo? firstError = null;
        try { _errors.OnCompleted(); }
        catch (Exception error) { firstError = ExceptionDispatchInfo.Capture(error); }
        try { _errors.Dispose(); }
        catch (Exception error) { firstError ??= ExceptionDispatchInfo.Capture(error); }
        firstError?.Throw();
    }
}

File: langs/python/src/vmx/commands/confirmation_decorator_command.py
"""ConfirmationDecoratorCommand — gates execution on an async confirm delegate.

See spec/04-commands.md §Decorators and ADR-0012.
"""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from concurrent.futures import Future
from typing import Any

from reactivex import Observable
from reactivex import operators as ops
from reactivex.subject import Subject

from vmx._asyncio_runner import submit_background
from vmx.commands.protocols import Command


class ConfirmationDecoratorCommand:
    """Wraps a single inner command, prompting an async ``confirm`` delegate
    before invoking. UI-agnostic per ADR-0012.
    """

    def __init__(
        self,
        inner: Command,
        confirm: Callable[[], Awaitable[bool]],
    ) -> None:
        self._inner = inner
        self._confirm = confirm
        self._disposed = False
        self._errors: Subject[BaseException] = Subject()

    @property
    def can_execute_changed(self) -> Observable[None]:
        return self._inner.can_execute_changed

    @property
    def errors(self) -> Observable[BaseException]:
        """Observable surfacing an error from the fire-and-forget :meth:`execute`.

        ``execute()`` schedules the confirm flow and returns immediately, so a
        rejecting confirm delegate or a throwing inner command cannot propagate
        to the caller the way the base command's task does. Instead of
        swallowing it, the error is emitted here (VMX-009). Await
        :meth:`execute_async` to observe it inline. Completes on :meth:`dispose`.
        """
        return self._errors.pipe(ops.as_observable())

    def can_execute(self, parameter: Any = None) -> bool:
        return self._inner.can_execute(parameter)

    def execute(self, parameter: Any = None) -> None:
        """Fire-and-forget. Schedules ``execute_async`` on the current event loop.

        A rejecting confirm delegate or a throwing inner command is routed to
        the :attr:`errors` channel instead of being swallowed (VMX-009).
        """
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            future = submit_background(self.execute_async(parameter))
            future.add_done_callback(self._on_done)
            return
        task = loop.create_task(self.execute_async(parameter))
        # Surface the failure on the error channel; skip cancelled tasks —
        # Task.exception() raises CancelledError for those, which would error
        # the loop's callback.
        task.add_done_callback(self._on_done)

    async def execute_async(self, parameter: Any = None) -> None:
        if not self.can_execute(parameter):
            return
        if await self._confirm():
            self._inner.execute(parameter)

    def _on_done(self, task: asyncio.Future[None] | Future[None]) -> None:
        if task.cancelled():
            return
        exc = task.exception()
        if exc is not None:
            self._emit_error(exc)

    def _emit_error(self, exc: BaseException) -> None:
        # The subject is completed+disposed on dispose(); a failure arriving
        # after dispose must not raise reactivex DisposedException.
        if self._disposed:
            return
        self._errors.on_next(exc)

    def dispose(self) -> None:
        """Mark the decorator as disposed and complete the :attr:`errors` channel.

        Idempotent. ``can_execute_changed`` delegates lazily to the inner
        command, so the decorator owns no other subscriptions to release.
        Provided for API symmetry with the C# IDisposable surface (see
        ``CompositeCommand.dispose``).
        """
        if self._disposed:
            return
        self._disposed = True
        self._errors.on_completed()
        self._errors.dispose()
